[DTensor] Add a private util for sharding tensor

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142288

Locally shards a full tensor based on indicated sharding arrangement, and returns a DTensor containing the local shard.

warning: This is a private API purposed to skip the communication otherwise required by `distribute_tensor`. It is only applicable to a case where all ranks have the same `full_tensor`. 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @tianyu-l @XilunWu

kwen2501 committed:
[DTensor] Add a private util for sharding tensor

[ghstack-poisoned]

wz337 reviewed (approved):
Stamped to unblock.

wz337 commented on torch/distributed/tensor/_api.py:
```diff
@@ -764,6 +764,51 @@ def distribute_tensor(
+    Locally shards a full tensor based on indicated sharding arrangement, and
+    returns a DTensor containing the local shard.
+
+    .. warning:: This is a private API purposed to skip the communication
```
Could we say `This is a private API that is subject to change. It is ...`

kwen2501 replied:
Thanks, added now.

wz337 commented on torch/distributed/tensor/_api.py:
```diff
@@ -764,6 +764,51 @@ def distribute_tensor(
+        for cur_shape, cur_offset in zip(shape, offset)
+    ]
+    local_tensor = full_tensor[slices]
+    return DTensor.from_local(local_tensor, device_mesh, placements)
```
You will need to pass the `shape` and `stride` for uneven tensor for `from_local`. Otherwise, the shape and stride would be inferred from the local tensor as it is uniformly distributed. See example: https://github.com/pytorch/pytorch/blob/main/torch/distributed/_state_dict_utils.py#L566-L572

kwen2501 replied:
Thanks for the heads-up. I don't have the shape or stride info, and this function assumes things are the same across ranks, so the infer would be fine.

kwen2501 committed:
Update on "[DTensor] Add a private util for sharding tensor"


Locally shards a full tensor based on indicated sharding arrangement, and returns a DTensor containing the local shard.

warning: This is a private API purposed to skip the communication otherwise required by `distribute_tensor`. It is only applicable to a case where all ranks have the same `full_tensor`. 

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o tianyu-l XilunWu

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge -f "CI was green; minor edits to comments"
